### PR TITLE
[fix] Reject SBA steps built on a failed Cholesky or on hidden points

### PR DESCRIPTION
**Pull request by slepichev**
GPUCholeskySolver ran cusolverDnSpotrf and went straight on to the triangular solve. cuSOLVER reports "not positive definite" through info, not through its return code, and info was memset and never read - so a partial factor was back-substituted and the pose step came back non-finite. That is what made assert(std::isfinite(predicted_relative_reduction)) unprovable and left it commented out behind a #warning. Read the status back before Spotrs overwrites it, expose solve_succeeded(), and have the bundler treat a failed factorization as a rejected step: damping is what fixes a non positive definite system. The CPU bundler never had this because it solves the same system with a thresholded JacobiSVD, which is why its copy of the assert survived. The assert is back.

The cost had a second way to be wrong. evaluate_cost_kernel skips observations behind the camera and they contribute nothing, so a state that hides all of them scores a perfect zero: an infeasible start was reported as converged, and a step that hid everything looked like an improvement. num_skipped was already computed for exactly this and thrown away - fold it in, the way the IMU bundler does, so such a state costs infinity and is refused.

Also stop compute_update from launching kernels with a zero grid or block dimension when nothing is left free to optimize.

Tests cover the vanished-cost precondition on predict_reduction and add the first end-to-end coverage of SchurComplementBundlerGpu, which had none.

Co-Authored-By: Claude Opus 5 (1M context) <noreply@anthropic.com>

## Summary by CodeRabbit

- **New Features**
  - Added solver status reporting for GPU-based optimization operations.
  - GPU bundle adjustment now handles skipped observations and failed optimization steps more reliably.

- **Bug Fixes**
  - Improved cost evaluation for infeasible states.
  - Prevented invalid optimization updates from being accepted.
  - Added validation for empty input data.

- **Tests**
  - Added GPU solver convergence and edge-case coverage.
  - Expanded Dockerized test execution to include all tests.

### CODE REVIEW

File: libs/cuda_modules/sba.h [hunk 1]
```diff
@@ -186,9 +189,14 @@ class GPUCholeskySolver {
   // Queues the Cholesky solve of A x = b on the lower triangle of A - it does not solve anything by
   // the time it returns. A and b are only read; x holds the system_order floats of the result once
   // the stream has run it. Nothing fails loudly: a matrix that is not positive definite still
-  // produces an x, and nothing here tells you that x is garbage.
+  // produces an x, and only solve_succeeded() tells you that x is garbage.
   void solve(const float* A, size_t A_pitch, const float* b, float* x, int system_order, cudaStream_t s);
```
Comment (vikuznetsov-nvidia) [inline libs/cuda_modules/sba.h:193]: Why not `bool solve()`? Does it work asynchronously?

File: libs/sba/schur_complement_bundler_gpu.cpp [hunk 2]
```diff
@@ -202,8 +226,7 @@ bool SchurComplementBundlerGpu::Impl::solve(BundleAdjustmentProblem& problem) {
     // cost = inf and current_cost = inf.
     // The only possible case is cost = inf (step leads outside of feasible region).
     assert(std::isfinite(current_cost));
-#warning Fix assert(std::isfinite(predicted_relative_reduction))
-    // FIXME assert(std::isfinite(predicted_relative_reduction));
+    assert(std::isfinite(predicted_relative_reduction));
```
Comment (vikuznetsov-nvidia) [inline libs/sba/schur_complement_bundler_gpu.cpp:229]: Nice

**Review by vikuznetsov-nvidia — DISMISSED** [review]

File: libs/cuda_modules/sba.h (hunk 1 above)
Comment (slepichev) [reply libs/cuda_modules/sba.h:193]: Yes it is.

**Review by vikuznetsov-nvidia — APPROVED** [review]

**Review by vikuznetsov-nvidia — APPROVED** [review]
